zebra-consensus/src/checkpoint.rs

```diff
@@ -485,40 +494,41 @@ where
         crate::block::check::difficulty_is_valid(&block.header, self.network, &height, &hash)?;
         crate::block::check::equihash_solution_is_valid(&block.header)?;
 
-        let transaction_hashes = block
-            .transactions
-            .iter()
-            .map(|tx| tx.hash())
-            .collect::<Vec<_>>();
+        // don't do precalculation until the block passes basic difficulty checks
+        let block = FinalizedBlock::with_hash_and_height(block, hash, height);
 
-        crate::block::check::merkle_root_validity(self.network, block, &transaction_hashes)?;
+        crate::block::check::merkle_root_validity(
+            self.network,
+            &block.block,
+            &block.transaction_hashes,
+        )?;
 
-        Ok(height)
+        Ok(block)
     }
 
-    /// Queue `block` for verification, and return the `Receiver` for the
-    /// block's verification result.
+    /// Queue `block` for verification.
+    /// On success, returns a [`RequestBlock`] containing the block,
```

Comment:
Nit: Maybe add a newline so that the first line becomes the documentation summary, and the return value description ends up in the details section?

```suggestion
    /// Queue `block` for verification.
    ///
    /// On success, returns a [`RequestBlock`] containing the block,
```